src/test/lib/scheduler/operator_task_test.cpp

```diff
@@ -107,4 +109,56 @@ TEST_F(OperatorTaskTest, MakeDiamondShape) {
     // We don't have to wait here, because we are running the task tests without a scheduler
   }
 }
+
+TEST_F(OperatorTaskTest, ConcurrentTaskReusage) {
+  // Operators reuse the created OperatorTasks when the tasks are still available. This may happen concurrently, e.g.,
+  // when an uncorrelated subquery is used in multiple TableScans. This test ensures that concurrently creating/reusing
+  // tasks from the same operator is thread-safe and does not lead to segmentation faults or tasks waiting forever to
+  // finish.
+  Hyrise::get().set_scheduler(std::make_shared<NodeQueueScheduler>());
+
+  const auto get_table = std::make_shared<GetTable>("table_b");
+  const auto scan_predicate = greater_than_(pqp_column_(ColumnID{0}, DataType::Int, false, "a"), 100);
+  const auto table_scan = std::make_shared<TableScan>(get_table, scan_predicate);
+
+  const auto num_threads = 100;
+  const auto iterations_per_thread = 50;
+  auto thread_futures = std::vector<std::future<void>>{};
+  thread_futures.reserve(num_threads);
+  auto threads = std::vector<std::thread>{};
+  threads.reserve(num_threads);
+  auto successful_executions = std::atomic_size_t{0};
+
+  for (auto thread_id = 0; thread_id < num_threads; ++thread_id) {
+    auto task = std::packaged_task<void()>{[&]() {
+      for (auto iteration = 0; iteration < iterations_per_thread; ++iteration) {
+        const auto& [tasks, _] = OperatorTask::make_tasks_from_operator(table_scan);
+        Hyrise::get().scheduler()->schedule_and_wait_for_tasks(tasks);
```

Comment:
Can you help the test reader here a little bit? Just a short comment, where the a task is shared and where we could run into problems if we don't wouldn't properly handle tasks?

Reply:
The comment is of course only for new students. Absolutely not for me.

Reply:
Maybe something like the general idea. If I understand it correctly, we will spawn 100 threads that all execute the same operator several times. As the operator task can be shared, it might happen that a lot of threads will re-use the already created task, right?
So we might have dozens of executions and hundreds of reuses?

Reply:
> If I understand it correctly, we will spawn 100 threads that all execute the same operator several times. As the operator task can be shared, it might happen that a lot of threads will re-use the already created task, right?
So we might have dozens of executions and hundreds of reuses?

Exactly. This test indeed crashes (not always, but often) when one of the two fixes in this PR, as there are multiple issues in the current version:
- Gap between weak pointer expiration check an locking in `AbstractOperator::get_or_create_operator_task`: when we have a bunch of concurrent threads, it is likely that this segfaults as at least one thread will peak in the expiration and get a `nullptr` rather than a pointer to a task (reusage fails). However, something has to be executed somewhen.
- Concurrent read/write of `AbstractTask::_successors`: Likely fails when an `OperatorTask` was created by one thread and is returned to another thread by the above mentioned method (a.k.a. reusage). Similarly, the more threads, the more likely we fail.
- Gap between duplicate checking and adding successor in the same method: Leads to deadlocks and again, the more threads, the better (reusage works, but consumer registration is incorrect).

At least for the first issue, we have to actually execute and destruct a task. For each one, many executions and parallel task creations help to trigger failures.

As a comment, is the following ok for the overall test?
```
// Operators reuse the created OperatorTasks when the tasks are still available. Requesting the tasks may happen con-
// currently, e.g., when an uncorrelated subquery is used in multiple TableScans (see #2520). This test ensures that
// concurrently creating/reusing tasks from the same operator is thread-safe and does not lead to segmentation faults
// or tasks waiting forever to finish. This test addresses multiple issues:
// - Operators store a weak pointer to an OperatorTask. If the task has been destructed, we must ensure to create the
//   task again instead of returning nullptr.
// - An AbstractTask stores its succeeding tasks in a vector. If different threads concurrently access this vector, we
//   must ensure that (i) parallel access is safe (e.g., via locking), and (ii) this vector does not contain dupli-
//   cates, which would lead to deadlocks.
// We test both requirements by letting many threads create and execute tasks from the same root operator concurrent-
// ly. Thus, we concurrently access a task's successors and likely have expiring tasks referenced by operators.
```

Reply:
Sounds good to me. The only minor comment would be on hyphenation. I think we don't do it anywhere else. I wouldn't vote against it, but it does not look familiar to me. :)

Reply:
Feel free to resolve when this comment is in the code (if it isn't already).

Reply:
Never thought about hyphenation tbh, and do not have a string opinion. Pinging @mweisgut here. Maybe sth we want to explicitly mention in the style guide?